tc/core/tc_executor_api_rfc.h

```diff
@@ -0,0 +1,138 @@
+ * implement the uncheckedRun and setRuntimeCompiledFunction.
+ */
+template <typename Backend>
+class TcExecutor {
+ protected:
+  TcExecutor(
+      lang::TreeRef tcDefinition,
+      const std::vector<const DLConstTensor*>& inputs);
+
+ public:
+  virtual ~TcExecutor();
+
+  TcExecutor(TcExecutor&&) = delete;
+  TcExecutor& operator=(TcExecutor&&) = delete;
+  TcExecutor(const TcExecutor&) = delete;
+  TcExecutor& operator=(const TcExecutor&) = delete;
+
+  /// Run can be called multiple times given a compilation, inputs are allowed
+  /// to change in that their data pointer is allowed to change.
+  /// Sizes and strides must remain constant otherwise this is an error.
+  /// The only thing that is allowed to change across runs is the input
+  /// and output pointers base address.
+  /// It is the caller's responsibility to ensure proper non-aliasing (or
+  /// advanced aliasing) properties of the input and output tensors.
+  /// if profile is set the kernel runtime (microseconds) is returned.
+  /// \returns profiling information (see: tc/core/utils/time.h)
+  ProfilingInfo run(
+      const std::vector<const DLConstTensor*>& inputs,
+      const std::vector<const DLTensor*>& outputs,
+      bool profileKernel = false) const;
+
+  /// It is necessary to clear the RTC manually because it can throw and we
+  /// can't have that in the destructor.
+  inline void clearRuntimeCompiledFunction();
+
+  /*****************************************************************************
+   * The following pure virtual methods are the minimal set of required
+   * backend-dependent functions.
+   ****************************************************************************/
+  /// This is the "low-latency" mode in which we just propagate raw pointers to
+  /// data in the address space where kernel is executed.
+  /// No tensor-related information can be checked so it is the user's
+  /// responsibility to ensure that shapes and strides match. If the user
+  /// doesn't then segfault will likely occur.
+  virtual Duration uncheckedRun(
+      const std::vector<const void*>& inputs,
+      const std::vector<void*>& outputs,
+      bool profileKernel) const = 0;
+
+ protected:
+  /// setupRuntimeCompiledFunction takes a compilationResult (returned by
+  /// the specialized TC mapper) and:
+  ///   a. saves backend-specific information required at runtime
+  ///     (e.g. grid and block sizes for the CUDA backend)
+  ///   b. creates the in-memory object resulting from JIT-compilation that
+  ///      implements the TC
+  virtual void setupRuntimeCompiledFunction(
+      const typename Backend::CompilationResultType& compilationResult) = 0;
+
+ protected:
+  /// Used to check proper metadata when calling run
+  std::vector<TensorInfo> inputsInfo_;
+  std::vector<TensorInfo> outputsInfo_;
+  tc2halide::HalideComponents halideComponents_;
+
+  /// The following are initialized as a result of compilation with the TcMapper
+  std::vector<int> parameters_;
```

Comment:
we moved to using `unordered_map<string, int>` elsewhere for parameters, maybe do this here too?

Reply:
Position is important (passed in order to rtc function call) and it is not necessarily ordered by name (so it can't be just a map).  And I'd claim we don't want to be in the business of reordering at runtime.